Gremlin Go MS5 - Feature Complete

lyndonbauto opened the pull request:
This has the changes for Milestone 5 of the gremlin-go driver. I would like to thank @xiazcy, @simonz-bq, @L0Lmaker, @vkagamlyk, @spmallette, and @krlawrence for continued help they have all been putting into this.

We have all tests passing in the Cucumber test suite, have integrated the driver into the TinkerPop build infrastructure and have added all remaining features to the driver at this point.

From here we will be working on enhancing documentation, adding performance testing, and bug fixes to any issues which are exposed.

Pull request comments will be addressed as part of the next milestone.

This pull request will remain open for at least a week, so the community has a chance to look at it and critique it.

spmallette commented on gremlin-console/bin/gremlin.sh:
```diff
@@ -1 +1,21 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
```
dont think this needs a license - it's just a symlink. wondering if it actually breaks the symlink to add that? in any case, if it's being recognized as text in a windows build and failing rat for some reason, maybe just add an exclusion for it.

spmallette commented on gremlin-go/Dockerfile:
```diff
@@ -0,0 +1,39 @@
+# under the License.
+
+ARG GREMLIN_SERVER_VERSION=${version:-3.5.4}
+FROM tinkerpop/gremlin-server:${GREMLIN_SERVER_VERSION}-SNAPSHOT
```
i'm not sure i understand how this works as docker is still something i'm learning. i understand what an `ARG` is but i don't follow what the implications are of hardcoding to 3.5.4 and then appending "-SNAPSHOT". what happens when we are doing build artifacts on 3.5.4 and we are no longer on "-SNAPSHOT"? Isn't there a nice way to gather the version from the `pom.xml` and pass it through so that all of this stays dynamic?

spmallette commented on gremlin-go/Dockerfile:
```diff
@@ -0,0 +1,39 @@
+RUN chmod 755 /opt/docker-entrypoint.sh
+
+# Setting to 3.5.3 does gives error even when using outside of Docker.
+ENV NEO4J_VERSION=${version:-3.5.2}
```
Unless I'm misunderstanding this section, pinning to a version is problematic especially to an older version. The version should come from the `pom.xml` so that it stays current. Is it possible there is an error because `neo4j-gremlin` needs to be explicitly built for docker's purposes?

spmallette commented on gremlin-go/docker/gremlin-server-integration-secure.yaml:
```diff
@@ -0,0 +1,74 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
we had managed to reduce the number of copies of these gremlin server config files down to:

* the template one in `gremlin-server/src/test/resources/org/apache/tinkerpp/gremlin/server` - this one is dynamically reconfigured by the groovy server start scripts to produce the three server variations that are tested
* the three configuration files that statically represent the three server variations described above and are stored in `docker/gremlin-server/`

How do we avoid duplicating these here in gremlin-go? and, if gremlin-go is meant to be the model for other GLVs then i think it would be worth including the third configuration even if we dont' support kerberos here.

divijvaidya commented on gremlin-go/driver/connectionPool.go:
```diff
@@ -0,0 +1,147 @@
+	loadBalanceLock        sync.Mutex
+}
+
+func (pool *loadBalancingPool) close() {
```
What happens when we call close() while a `createConnection()` is in progress? I think we would close the registered connections and would never close the connection that will be added to the pool after `createConnection()` is complete.

divijvaidya commented on gremlin-go/driver/connectionPool.go:
```diff
@@ -0,0 +1,147 @@
+func newLoadBalancingPool(url string, logHandler *logHandler, connSettings *connectionSettings, newConnectionThreshold int,
+	maximumConcurrentConnections int) (connectionPool, error) {
+	pool := make([]*connection, 0, maximumConcurrentConnections)
+	initialConnection, err := createConnection(url, logHandler, connSettings)
+	if err != nil {
+		return nil, err
+	}
+	pool = append(pool, initialConnection)
```
Code duplication. Refactor `newConnection()` to work for this use case as well instead?

divijvaidya commented on gremlin-go/driver/connectionPool.go:
```diff
@@ -0,0 +1,147 @@
+func newLoadBalancingPool(url string, logHandler *logHandler, connSettings *connectionSettings, newConnectionThreshold int,
+	maximumConcurrentConnections int) (connectionPool, error) {
+	pool := make([]*connection, 0, maximumConcurrentConnections)
+	initialConnection, err := createConnection(url, logHandler, connSettings)
```
The number of initial connections per pool should be configurable. We have faced situations with Java client where users complained about the fact that "lazy initialization" leads to higher latency for first few requests to the server.

divijvaidya commented on gremlin-go/driver/connectionPool.go:
```diff
@@ -0,0 +1,147 @@
+	}
+}
+
+func (pool *loadBalancingPool) write(request *request) (ResultSet, error) {
```
what happens when write() is called when another thread is executing the close() method?

divijvaidya commented on gremlin-go/driver/connectionPool.go:
```diff
@@ -0,0 +1,147 @@
+	defer pool.loadBalanceLock.Unlock()
+	if len(pool.connections) == 0 {
+		return pool.newConnection()
+	} else {
```
This else block is not required since there is no code to execute outside the else block.

divijvaidya commented on gremlin-go/driver/connection.go:
```diff
@@ -52,7 +60,7 @@ func (connection *connection) errorCallback() {
 	if connection.state != established {
-		return errors.New("cannot close connection that has already been closed or has not been connected")
+		return newError(err0101ConnectionCloseError)
 	}
```
We need to introduce a 'closing' state here. For the same connection, this method could be called simultaneously from two threads (e.g. one from go routine inside pool.write and another from pool.close). Only one should be allowed to 'close'.

Also please add method doc which mentions whether a function is idempotent or not and also whether it is thread safe or not.

divijvaidya commented on gremlin-go/driver/connectionPool.go:
```diff
@@ -0,0 +1,147 @@
+		for _, connection := range pool.connections {
+			// Purge dead connections from pool
+			if connection.state == established {
+				// Close and purge connections from pool if there is more than one being unused
```
I am not very sure about this logic. Connection setup is quite expensive due to SSL handshake, authn and authz happening with every setup. It is very likely that a connection may be intermittently un-used and this logic will end up creating and destroying connections all the time!

Alternatively, we can do the following:
1. Close "idle" connections based on a percentage configured by the user.
2. Do not close the idle connections at all. Keeping connections alive is not expensive on the server.

Please add pros/cons about this in the design and send an email to the community mailing list for a design review/discussion.

divijvaidya commented on gremlin-go/driver/connectionPool.go:
```diff
@@ -0,0 +1,147 @@
+					leastUsed = connection
+				}
+			} else {
+				pool.logHandler.log(Debug, purgingDeadConnection)
```
should there be a function call here to purge connections?

divijvaidya commented on gremlin-go/driver/connectionPool.go:
```diff
@@ -0,0 +1,147 @@
+			}
+		}
+
+		// Deallocate truncated dead connections to prevent memory leak
```
perhaps an attempt to gracefully close them first?

divijvaidya commented on gremlin-go/driver/connectionPool.go:
```diff
@@ -0,0 +1,147 @@
+		validIndex := 0
+		for _, connection := range pool.connections {
+			// Purge dead connections from pool
+			if connection.state == established {
```
this might end up purging "initialized" state connections too.

lyndonbauto commented:
@divijvaidya Thank you for looking over the code! I am going to be addressing these comments as part of the MS6 delivery as discussed. The feedback is really appreciated :)

lyndonbauto commented:
Thanks for reviewing @spmallette and @divijvaidya. 

The comments @divijvaidya made are being actively worked on (most have been handled) in another branch.

The comments @spmallette made about the symlink file will be handled here before merging, the comments about docker will be handled separately in this ticket: https://issues.apache.org/jira/browse/TINKERPOP-2737.
